### PR TITLE
V2 Cookbook Manifests

### PR DESCRIPTION
**Pull request by thommay**
Use v2 cookbook manifests, and remove deprecated metadata fields.

### CODE REVIEW

**Comment by thommay** [comment]
This is bloody close to working, @chef/maintainers . Pre-review appreciated (mostly of the ripping out of almost all of CookbookVersion and moving that to the manifest).

**Comment by thommay** [comment]
@coderanger this clobbers the poise tests, which seemed to be the only user of `attribute_files` in existence. Replacement is `files_for("attributes")`…

File: lib/chef/chef_fs/file_system/repository/chef_repository_file_system_cookbook_dir.rb [hunk 1]
```diff
@@ -97,9 +97,9 @@ def children
           end
 
           def can_have_child?(name, is_dir)
-            if is_dir
+            if is_dir && !%w{ root_files .. . }.include?(name)
```
Comment (danielsdeleo) [inline lib/chef/chef_fs/file_system/repository/chef_repository_file_system_cookbook_dir.rb]: Should this be something like `!(root_files + %w{ .. . }).include?(name)` ? Looks like that array literal evaluates to `['root_files', '..', '.']` ?

File: lib/chef/cookbook/metadata.rb [hunk 2]
```diff
@@ -726,7 +617,7 @@ def method_missing(method, *args, &block)
         if block_given?
           super
         else
-          Chef::Log.debug "ignoring method #{method} on cookbook with name #{name}, possible typo or future metadata?"
+          Chef::Log.debug "ignoring method #{method} on cookbook with name #{name}, possible typo or the ghosts of metadata past or future?"
```
Comment (danielsdeleo) [inline lib/chef/cookbook/metadata.rb]: This puts a smile on my face, but maybe something more informative would be helpful.

File: lib/chef/chef_fs/file_system/repository/chef_repository_file_system_cookbook_dir.rb (hunk 1 above)
Comment (thommay) [inline lib/chef/chef_fs/file_system/repository/chef_repository_file_system_cookbook_dir.rb]: nope, name will actually be `root_files` here.

File: lib/chef/cookbook/metadata.rb (hunk 2 above)
Comment (thommay) [inline lib/chef/cookbook/metadata.rb]: yeah, i mostly agree, but it's a ray of sunshine for me amidst a dark sea of cheffs

**Comment by thommay** [comment]
OK, barring bundler insanity, this is ready to go. Note that this must be merged with chef/chef-zero#250 and chef/chef-server#1034 once that's finished.

Please review, @chef/maintainers

**Comment by coderanger** [comment]
@thom I'll get my side of things cleaned up ASAP.

**Comment by lamont-granquist** [comment]
should squash the commits

File: Gemfile.lock [hunk 3]
```diff
@@ -142,46 +154,6 @@ PATH
       specinfra (~> 2.10)
       syslog-logger (~> 1.6)
       uuidtools (~> 2.1.5)
-    chef (13.0.79-universal-mingw32)
```
Comment (lamont-granquist) [inline Gemfile.lock]: need to regen this using the rake task

Comment (thommay) [inline Gemfile.lock]: the rake task assploads :(

File: lib/chef/cookbook_version.rb [hunk 4]
```diff
@@ -235,41 +194,7 @@ def load_recipe(recipe_name, run_context)
...
-    def replace_segment_filenames(segment, filenames)
-      case segment.to_sym
-      when :recipes
-        self.recipe_filenames = filenames
-      when :attributes
-        self.attribute_filenames = filenames
-      else
-        segment_filenames(segment).replace(filenames)
-      end
+      files_for(segment)
```
Comment (coderanger) [inline lib/chef/cookbook_version.rb]: This is returning manifest entries, not actual filenames. To preserve some level of compat it should probably be

```ruby
files_for(segment).map {|file| file["full_path"] }
```

However I put a workaround in my code so if we want to change the API I can just leave that in place.

**Comment by coderanger** [comment]
New Poise and Halite pushed with feature-flagged forward compat for this branch.

File: Gemfile.lock (hunk 3 above)
Comment (lamont-granquist) [inline Gemfile.lock]: hrm, `rake dependencies:update_gemfile_lock` or `rake dependencies`?  anyway send me the assplosion and i could look at it...

**Pull request merged**